File: test/serial/newlib_full.py
import argparse
import errno
import os

from metal.serial import Engine, Exit
from subprocess import PIPE, Popen

from metal.serial.generate import generate

from metal.serial.newlib import build_newlib_hook

parser = argparse.ArgumentParser()

parser.add_argument('binary',           help='The binary that runs on target')
parser.add_argument('-S', '--source-dir',  required=True, help='The root of the source directory')
parser.add_argument('-I', '--include', nargs='+', help="Include folders for the preprocessor", default=[])
parser.add_argument('-D', '--define', nargs='+', help="Defines for the preprocessor", default=[])

args = parser.parse_args()

serial_info = generate(args.binary, args.define, args.include)

p = Popen(args.binary, stdin=PIPE, stdout=PIPE, close_fds=True)

class os_stub:
    stat_result = os.stat_result
    O_WRONLY = 1
    O_APPEND = 2
    O_CREAT  = 4
    O_EXCL   = 5
    O_NOINHERIT = 6
    O_RANDOM = 7
    O_RDONLY = 8
    O_RDWR   = 9
    O_SEQUENTIAL = 10
    O_NOCTTY = 11
    O_NONBLOCK = 12
    O_SYNC = 13
    O_CLOEXEC = 14

    O_DIRECT = 15
    O_DIRECTORY = 16
    O_DSYNC = 17

    O_NOATIME = 18
    O_NDELAY = 19
    O_PATH = 20
    O_TRUNC = 21

    O_READ = 22
    O_WRITE = 23
    O_RDWR = 24

    SEEK_SET = 42
    SEEK_CUR = 43
    SEEK_END = 44

    S_IREAD   = 1
    S_IWRITE  = 2
    S_IREAD   = 3
    S_IWRITE  = 4
    S_IRWXU   = 5
    S_IRUSR   = 6
    S_IWUSR   = 7
    S_IXUSR   = 8
    S_IRWXG   = 9
    S_IRGRP   = 10
    S_IWGRP   = 11
    S_IXGRP   = 12
    S_IRWXO   = 13
    S_IROTH   = 14
    S_IWOTH   = 15
    S_IXOTH   = 16
    S_ISUID   = 17
    S_ISGID   = 18
    S_ISVTX   = 19
    S_IEXEC   = 10
    S_ENFMT   = 21
    S_IFMT    = 22
    S_IFDIR   = 23
    S_IFCHR   = 24
    S_IFBLK   = 25
    S_IFREG   = 26
    S_IFLNK   = 27
    S_IFSOCK  = 28
    S_IFIFO   = 29

    def __init__(self):
        self.write_cnt = 0
        self.open_cnt = 0
        self.close_cnt = 0
        self.fds = [1, 2]

    def write(self, fd, msg):
        assert fd in self.fds
        if fd == 1:
            assert msg == b"Writing stdout\n"
        if fd == 2:
            assert msg == b"Writing stderr\n"
        if fd == 3:
            assert msg == b"Writing to fd_\n"

        self.write_cnt = self.write_cnt + 1
        return len(msg)

    def open(self, filename, flags, mode):
        assert filename == "test-file"
        assert flags & os_stub.O_RDWR
        assert mode == 0
        self.open_cnt  = self.open_cnt + 1

        self.fds.append(3)
        return 3

    def close(self, fd):
        assert fd == 3
        self.close_cnt = self.close_cnt + 1

    def read(self, fd, n):
        assert n == 5
        if fd == 0:
            return b'stdin'
        elif fd == 3:
            return b'testf'
        else:
            assert False

    def stat(self, name):
        assert name.endswith('newlib_full.c')
        raise OSError(errno.ENOENT)

    def fstat(self, name):
        st = os.stat_result((
            3, 2, 1, 4, 5, 6, 1024, 900.000000800, 3000, 5000
            ))

        return st

    def isatty(self, fd):
        if fd == 1:
            return False
        if fd == 2:
            return True

    def link(self, existing, new):
        assert existing.endswith('syscalls.h')
        assert new.endswith('lib/syscalls.c')
        return 0

    def symlink(self, existing, new):
        assert existing.endswith("doesn't exist")
        assert new.endswith('lib/syscalls.h')
        raise OSError(errno.ENOENT)

    def unlink(self, existing):
        assert existing.endswith('syscalls.h')
        return 0

    def lseek(self, fd, ptr, dir):
        assert fd in [0, 3]
        assert ptr == 0

        if dir == os_stub.SEEK_CUR:
            return 0
        if dir == os_stub.SEEK_END:
            return 5




oss = os_stub()
engine = Engine(input=p.stdout, output=p.stdin, serial_info=serial_info,
                macro_hooks=[Exit, build_newlib_hook(oss)])

assert engine.init_marker.file.endswith('newlib_full.c')
assert engine.run() == 0
assert oss.write_cnt == 3
assert oss.open_cnt == 1

File: test/serial/newlib_full.c
/**
 * @file   newlib_blocked.c
 * @date   17.06.2020
 * @author Klemens D. Morgenstern
 *
 */

// !!! This test is line sensitive, do NOT change lines around.

#include <metal/serial/core.h>
#include <metal/serial/syscalls.h>
#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <assert.h>

int _open (char* file, int flags, int mode);
int _read (int file, char* ptr, int len);
int _write(int file, char* ptr, int len);
int _lseek(int file, int ptr, int dir);
int _close(int);
int _isatty(int file);
int _stat(const char* file, struct stat* st);
int _fstat(int fd, struct stat* st);

int _link(char* existing, char* _new);
int _symlink(char* existing, char* _new);
int _unlink(char* existing);

char metal_serial_read() { return fgetc(stdin);}
void metal_serial_write(char c) { fputc(c, stdout);}


int main(int argc, char ** args)
{
    freopen(NULL, "rb", stdin);
    freopen(NULL, "wb", stdout);

    setvbuf(stdin, NULL, _IONBF, 0);
    setvbuf(stdout, NULL, _IONBF, 0);

    METAL_SERIAL_INIT();

    int res = 0;

    struct stat st;
    assert(_stat(__FILE__, &st) != 0);
    assert(errno == 2);

    errno = 0;

    int len = sizeof("Writing stdout\n") - 1;
    assert(_write(STDOUT_FILENO, "Writing stdout\n", len) == len);
    assert(errno == 0);

    assert(_write(STDERR_FILENO, "Writing stderr\n", len) ==len);
    assert(errno == 0);

    errno = 0;
    assert(_isatty(STDOUT_FILENO) == 0);
    assert(errno == 0);

    assert(_isatty(STDERR_FILENO) != 0);
    assert(errno == 0);

    char buf[10];
    assert(_read(STDIN_FILENO, buf, 5) == 5);
    assert(errno == 0);
    assert(buf[0] == 's');
    assert(buf[1] == 't');
    assert(buf[2] == 'd');
    assert(buf[3] == 'i');
    assert(buf[4] == 'n');


    int fd = _open("test-file", O_RDWR, 0);
    assert(fd == 3);
    assert(errno == 0);

    assert(_write(fd, "Writing to fd_\n", len) == len);
    assert(errno == 0);


    assert(_read(fd, buf, 5) == 5);
    assert(errno == 0);
    assert(buf[0] == 't');
    assert(buf[1] == 'e');
    assert(buf[2] == 's');
    assert(buf[3] == 't');
    assert(buf[4] == 'f');


    assert(_fstat(fd, &st) == 0);
    assert(errno == 0);

    assert(st.st_dev == 1);
    assert(st.st_ino == 2);
    assert(st.st_nlink == 4);
    assert(st.st_uid == 5);
    assert(st.st_gid == 6);

    assert(st.st_size == 1024);
    assert(st.st_atim.tv_sec  == 900);
    assert(st.st_atim.tv_nsec == 800);
    assert(st.st_mtim.tv_sec  == 3000);
    assert(st.st_ctim.tv_sec  == 5000);

    errno = 0;
    assert(_close(fd) == 0);
    assert(errno == 0);

    assert(_link("syscalls.h", "lib/syscalls.c") == 0);
    assert(errno == 0);

    assert(_unlink("syscalls.h") == 0);
    assert(errno == 0);

    assert(_symlink("doesn't exist", "lib/syscalls.h") != 0);
    assert(errno == ENOENT);

    METAL_SERIAL_EXIT(res);

    return res;
}